cishouseholds/impute.py

```diff
@@ -83,13 +83,52 @@ def fill_forward_work_columns(
         .orderBy(F.col(visit_date_column).asc())
         .rowsBetween(Window.unboundedPreceding, Window.currentRow)
     )
-    for col in fill_forward_columns:
-        df = df.withColumn(
-            col,
-            F.when(
-                F.col(main_job_changed_column) != "Yes", F.last(F.col(col), ignorenulls=True).over(window)
-            ).otherwise(F.col(col)),
-        )
+    # TODO: find if theres a yes in main_job_changed_column
+    df = df.withColumn("FLAG_not-yes", F.when(F.col(main_job_changed_column) == "Yes", 1).otherwise(0)).withColumn(
+        "FLAG_not-yes", F.sum("FLAG_not-yes").over(window)
+    )
+    df = df.withColumn("FLAG_fillforwards_yes", F.when((F.col(main_job_changed_column) == "Yes"), 1))
+    df = df.withColumn(
+        "FLAG_fillforwards_no",
+        F.when(
+            (F.col(main_job_changed_column) != "Yes") | (F.col(main_job_changed_column).isNull()),
+            F.dense_rank().over(window),
+        ),
+    )
+    df = df.withColumn(
+        "FLAG_fillforwards_no",
+        F.when(F.col("FLAG_fillforwards_no") != 1, None).otherwise(F.col("FLAG_fillforwards_no")),
+    )
+    df = df.withColumn("FILL_forward", F.coalesce(F.col("FLAG_fillforwards_yes"), F.col("FLAG_fillforwards_no"))).drop(
+        "FLAG_fillforwards_yes", "FLAG_fillforwards_no"
+    )
+
+    df_fill_forwards_from = (
+        df.where(F.col("FILL_forward") == 1)
```

Comment:
Instead of creating the flags, create a condition that is:
if =="Yes" OR it's the first row within the window

(F.col(main_job_changed_column) == "Yes") | (F.row_number().over(window) == 1)